sympy/solvers/solveset.py

```diff
@@ -2434,60 +2447,96 @@ def linear_coeffs(eq, *syms, **_kw):
+
+        >>> linear_coeffs(x*(y + 1) + 3, x, y)
+        Traceback (most recent call last):
+        ...
+        NonlinearError: nonlinear term: x*(y + 1)
+
+    An error is not raised when dependent symbols are passed, however:
+
+        >>> linear_coeffs(x + 2*x**2, x, x**2)
+        [1, 2, 0]
+
+    To allow coefficients to contain symbol-dependent factors which are
+    not strictly one of the symbols, use keyword `strict=False`.
+    A NonlinearError will still be reported if a term contains two
+    literal symbols as factors:
+
+        >>> linear_coeffs(x*(y**2 + 1) + x*y, x, y, strict=False)
+        Traceback (most recent call last):
+        ...
+        NonlinearError: nonlinear term: x*y
+
+    But the error will no longer raise if the unexpanded expression
+    does not contain such a term:
+
+        >>> eq, v = x*(y**2 + 1) + 3, [x, y]
+        >>> linear_coeffs(eq, *v, strict=False)
+        [y**2 + 1, 0, 3]
+
+    This is likely most useful when dealing with functions and their
+    derivatives since a derivative "has" the function in its expression
+    but it might make sense to treat them as being independent.
+
+    Regardless of the setting of `strict`, it should always be
+    true that the result will reconstruct an expression that is
+    equivalent to the input expression.
+
+        >>> from math import prod
+        >>> r = _  # result from above
+        >>> eq.equals(r[-1] + sum([prod(i) for i in zip(v, r)]))
+        True
     """
-    d = defaultdict(list)
+    from sympy.polys.matrices.linsolve import _lin_eq2dict
     eq = _sympify(eq)
+    if len(syms) == 1 and iterable(syms[0]) and not isinstance(syms[0], Basic):
+        raise ValueError('pass unpacked symbols, *syms')
     symset = set(syms)
     if len(symset) != len(syms):
         raise ValueError('duplicate symbols given')
-    has = set(iterfreeargs(eq)) & symset
-    if not has:
-        return [S.Zero]*len(syms) + [eq]
-    c, terms = eq.as_coeff_add(*has)
-    d[0].extend(Add.make_args(c))
-    for t in terms:
-        m, f = t.as_coeff_mul(*has)
-        if len(f) != 1:
-            break
-        f = f[0]
-        if f in symset:
-            d[f].append(m)
-        elif f.is_Add:
-            d1 = linear_coeffs(f, *has, **{'dict': True})
-            d[0].append(m*d1.pop(0))
-            for xf, vf in d1.items():
-                d[xf].append(m*vf)
-        else:
-            break
-    else:
-        for k, v in d.items():
-            d[k] = Add(*v)
-        if not _kw:
-            return [d.get(s, S.Zero) for s in syms]+ [d[0]]
-        return d  # default is still list but this won't matter
-    raise NonlinearError('nonlinear term encountered: %s' % t)
+    try:
+        c, d = _lin_eq2dict(eq, symset, strict)
+    except PolyNonlinearError as err:
+        # try raise a helpful error
+        if strict:
+            try:
+                _linear_eq_to_dict([eq], syms, strict=False, _expand=False)
```

Comment:
We shouldn't call `_linear_eq_to_dict` again. Any code calling this that just wants to catch the exception and do something else would be slowed down.